### PR TITLE
Fix migration order v1.3

### LINKED ISSUE #3155: Migrations bug for upgrade from 1.2 to 1.3
- Gitea version (or commit ref): 1.3.1+1-g8a19c6b9
- Operating system: Ubuntu 16.04
- Database (use `[x]`):
  - [x] PostgreSQL
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Not relevant

## Description

I think I have discovered a serious migrations bug when upgrading from version 1.2.3 to version 1.3.1 today. It seems that several migrations were backported to the 1.2 branch out of order. Our Gitea installation had a database version of 42 before the upgrade and was upgraded to version 49. However, due to the ordering of the migrations, three of them were applied twice and three were not applied at all.

I becomes clearer if looking at the diff of `models/migrations/migrations.go` from the 1.2 to the 1.3 branch:

```diff
@@ -126,12 +126,26 @@ var migrations = []Migration{
        NewMigration("unescape user full names", unescapeUserFullNames),
        // v38 -> v39
        NewMigration("remove commits and settings unit types", removeCommitsUnitType),
-       // v43 -> v44
-       NewMigration("fix protected branch can push value to false", fixProtectedBranchCanPushValue),
+       // v39 -> v40
+       NewMigration("adds time tracking and stopwatches", addTimetracking),
+       // v40 -> v41
+       NewMigration("migrate protected branch struct", migrateProtectedBranchStruct),
+       // v41 -> v42
+       NewMigration("add default value to user prohibit_login", addDefaultValueToUserProhibitLogin),
        // v42 -> v43
        NewMigration("add tags to releases and sync existing repositories", releaseAddColumnIsTagAndSyncTags),
+       // v43 -> v44
+       NewMigration("fix protected branch can push value to false", fixProtectedBranchCanPushValue),
        // v44 -> v45
        NewMigration("remove duplicate unit types", removeDuplicateUnitTypes),
+       // v45 -> v46
+       NewMigration("remove index column from repo_unit table", removeIndexColumnFromRepoUnitTable),
+       // v46 -> v47
+       NewMigration("remove organization watch repositories", removeOrganizationWatchRepo),
+       // v47 -> v48
+       NewMigration("add deleted branches", addDeletedBranch),
+       // v48 -> v49
+       NewMigration("add repo indexer status", addRepoIndexerStatus),
 }
 
 // Migrate database to current version
```

As can be seen, v43, v42 and v44 were applied (in that order) in the 1.2 branch (but with database versions of 40, 41 and 42). v39, v40 and v41 were then added before the existing migrations in the 1.3 branch. This led to v39, v40 and v41 being skipped, while v42, v43 and v44 have been applied a second time:

```
2017/12/11 16:13:30 [I] Migration: add tags to releases and sync existing repositories
2017/12/11 16:13:30 [I] Migration: fix protected branch can push value to false
2017/12/11 16:13:31 [I] Migration: remove duplicate unit types
2017/12/11 16:13:31 [I] Migration: remove index column from repo_unit table
2017/12/11 16:13:31 [I] Migration: remove organization watch repositories
2017/12/11 16:13:31 [I] Migration: add deleted branches
2017/12/11 16:13:31 [I] Migration: add repo indexer status
```

### PR DESCRIPTION
**Pull request by lafriks**
Specific to 1.3.x release when upgrading from 1.2.x
Fixes #3155 

PR to fix on master branch will follow in other PR so that there is no problems when upgrading from 1.2.x and 1.3.x to 1.4.0

### CODE REVIEW

**Comment by sapk** [comment]
LGTM

**Review by sapk — APPROVED** [review]

**Comment by lafriks** [comment]
Did file name renames to match real migration version, so diff does looks bigger than actual changes are because of that

File: models/migrations/migrations.go [hunk 1]
```diff
@@ -59,6 +59,10 @@ type Version struct {
 	Version int64
 }
 
+func emptyMigtation(x *xorm.Engine) error {
```
Comment (ethantkoenig) [inline models/migrations/migrations.go]: typo

**Review by ethantkoenig — COMMENTED** [review]
Are there any changes to the migrations (other than reordering/renaming)? It's hard to tell with the diff.

File: models/migrations/migrations.go (hunk 1 above)
Comment (lafriks) [reply models/migrations/migrations.go]: omg :) thanks :)

**Comment by lafriks** [comment]
@ethantkoenig Not really, just skipping column drop errors (in case migration is being run second time)

**Comment by lafriks** [comment]
@ethantkoenig fixed typo

**Review by ethantkoenig — APPROVED** [review]
Looks good, just one nit: could we add comments to where errors are ignored? It will help future readers understand what's going on, and may be useful if there turns out to be something wrong in this PR (hopefully not, but let's be safe).

**Comment by lafriks** [comment]
@ethantkoenig added comment before ignored error warn log message

**Comment by ethantkoenig** [comment]
~LGTM~ see below

File: models/migrations/v45.go [hunk 2]
```diff
@@ -18,6 +18,7 @@ func removeIndexColumnFromRepoUnitTable(x *xorm.Engine) (err error) {
 		log.Warn("Unable to drop columns in SQLite")
 	case setting.UseMySQL, setting.UsePostgreSQL, setting.UseMSSQL, setting.UseTiDB:
 		if _, err := x.Exec("ALTER TABLE repo_unit DROP COLUMN `index`"); err != nil {
+			// Ignoring this error in case we run this migration second time (after migration reordeting)
```
Comment (ethantkoenig) [inline models/migrations/v45.go]: We are not ignoring this error. Also typo (reordeting)

__EDIT__: I meant we are not *currently* ignornig this error; I did not mean that we *shouldn't* ignore this error. Sorry for any confusion

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

File: models/migrations/v45.go (hunk 2 above)
Comment (lafriks) [reply models/migrations/v45.go]: Then there is no way to fix this error

Comment (lafriks) [reply models/migrations/v45.go]: At least I don't see how we can fix to run 3 migrations again that were not run for all who upgraded from 1.2 -> 1.3 without running them for all and ignoring this error

Comment (ethantkoenig) [reply models/migrations/v45.go]: I'm not sure what you mean; can't we just replace the return with `log.Warn`?

**Review by ethantkoenig — APPROVED** [review]
LGTM - I don't want to block an important bug fix, so we can fix nits later

File: models/migrations/v45.go (hunk 2 above)
Comment (lafriks) [reply models/migrations/v45.go]: @ethantkoenig I already do log.Warn this error, see code below this comment. By ignoring I meant we do not fail migration by returning error

Comment (ethantkoenig) [reply models/migrations/v45.go]: line 22 (directly beneath this comment) is a `return` statement

Comment (lafriks) [reply models/migrations/v45.go]: Ooooh... Sry, just woke up and did not have my coffee yet :))) I thought you are saying this error can't be ignored but I just have added comment also to wrong file :)

Comment (lafriks) [reply models/migrations/v45.go]: I thought you are commenting on v51 migration where there is log.Warn and same comment

**Comment by lafriks** [comment]
@ethantkoenig fixed :)

**Comment by michaelkuhn** [comment]
Thanks for fixing this so quickly! One note, though: It might be a good idea to add a comment to `migrations.go` to never ever reorder migrations so this does not happen again in the future.

**Comment by lafriks** [comment]
Yes, I will add comment in master branch code

**Comment by moritz-h** [comment]
@lafriks Are there any corruptions to the database when I have already upgraded to 1.3 and if yes will this also be automatically fixed?
I'm currently on 1.3.0, which was upgraded from a fresh 1.2.3 installation and have the same log entries as in the issue report.

**Comment by lafriks** [comment]
@moritz-h there is just missing migrations so some features might not work as expected. Yes, 1.3.2 release will fix that.

**Comment by sapk** [comment]
@lafriks maybe we should add a deprecated message on 1.3.X releases messages when 1.3.2 is release ? To indicate the existing bug.

**Comment by lafriks** [comment]
All not latest minor versions should be considered deprecated by default :)

**Pull request merged**